Index: PythonDeviceHL.xtend
===================================================================
--- PythonDeviceHL.xtend	(revision 30002)
+++ PythonDeviceHL.xtend	(working copy)
@@ -141,6 +141,11 @@
     # ----------
 
 «cls.pythonAttributeDefinitions»
+    # -----
+    # Pipes
+    # -----
+
+«cls.pythonDevicePipes»
     # ---------------
     # General methods
     # ---------------
@@ -151,6 +156,11 @@
     # ------------------
 
 «cls.pythonAttributes»
+    # -------------
+    # Pipes methods
+    # -------------
+
+«cls.pythonPipes»
     # --------
     # Commands
     # --------
@@ -174,9 +184,10 @@
 from PyTango import DebugIt
 from PyTango.server import run
 from PyTango.server import Device, DeviceMeta
-from PyTango.server import attribute, command
+from PyTango.server import attribute, command, pipe
 from PyTango.server import class_property, device_property
-from PyTango import AttrQuality, AttrWriteType, DispLevel, DevState
+from PyTango import AttrQuality,DispLevel, DevState
+from PyTango import AttrWriteType, PipeWriteType
 «cls.inheritedAdditionalImportHL»
 # Additional import
 «IF cls.description.filestogenerate.toLowerCase.contains("protected regions")»«cls.protectedAreaHL("additionnal_import")»«ENDIF»
@@ -268,6 +279,16 @@
             dev.debug_stream("Details: " + traceback.format_exc())
     '''
 
+    //====================================================
+    //    Pipes
+    //====================================================
+    def pythonPipes(PogoDeviceClass cls)  '''
+«FOR pip: cls.pipes»
+«IF true»    «readPipeMethodHL(cls, pip)»«ENDIF»
+«IF pip.rwType.contains("WRITE")»    «writePipeMethodHL(cls, pip)»«ENDIF»
+«IF !pip.readExcludedStates.empty || !pip.writeExcludedStates.empty»    «pipeMethodStateMachineHL(cls, pip)»«ENDIF»
+«ENDFOR»
+    '''
     //    Commands
     //====================================================
     def pythonCommands(PogoDeviceClass cls)  '''
@@ -295,6 +316,16 @@
 «ENDFOR»
 «ENDIF»
     '''
+
+    //====================================================
+    //    Pipes
+    //====================================================
+    def pythonDevicePipes(PogoDeviceClass cls)'''
+«IF !cls.pipes.empty»
+«FOR pip : cls.pipes»    «pip.pythonPipeClassHL»
+«ENDFOR»
+«ENDIF»
+    '''
     
     //====================================================
     //    Main method
Index: PythonTypeDefinitions.java
===================================================================
--- PythonTypeDefinitions.java	(revision 30002)
+++ PythonTypeDefinitions.java	(working copy)
@@ -50,6 +50,7 @@
 import fr.esrf.tango.pogo.pogoDsl.DoubleType;
 import fr.esrf.tango.pogo.pogoDsl.DoubleVectorType;
 import fr.esrf.tango.pogo.pogoDsl.EncodedType;
+import fr.esrf.tango.pogo.pogoDsl.EnumType;
 import fr.esrf.tango.pogo.pogoDsl.FloatArrayType;
 import fr.esrf.tango.pogo.pogoDsl.FloatType;
 import fr.esrf.tango.pogo.pogoDsl.FloatVectorType;
@@ -202,6 +203,7 @@
 		if (type instanceof ULongArrayType)			return "('uint64')";
 		if (type instanceof DevIntType)				return "'DevInt'";
 		if (type instanceof EncodedType)			return "'bytearray'";
+		if (type instanceof EnumType)				return "'DevEnum'";
 		return "";
 	}
 
@@ -228,6 +230,7 @@
 		if (attr.getDataType() instanceof ULongType)			l_str = "'uint64'";
 		if (attr.getDataType() instanceof DevIntType)			l_str = "'DevInt'";
 		if (attr.getDataType() instanceof EncodedType)			l_str = "'bytearray'";
+		if (attr.getDataType() instanceof EnumType)				l_str = "'DevEnum'";
 		
 		if (attr.getAttType().equals("Spectrum"))
 		{
@@ -301,6 +304,7 @@
 		if (attr.getDataType() instanceof ULongType)			def_val =  "0";
 		if (attr.getDataType() instanceof DevIntType)			def_val =  "0";
 		if (attr.getDataType() instanceof EncodedType)			def_val =  "['', '']";
+		if (attr.getDataType() instanceof EnumType)				def_val =  "0";
 		
 		if (attr.getAttType().equals("Spectrum"))
 		{
@@ -498,6 +502,7 @@
 		if (type instanceof ULongArrayType)			return "return [0]";
 		if (type instanceof DevIntType)				return "return 0";
 		if (type instanceof EncodedType)			return "return \"\", \"\"";
+		if (type instanceof EnumType)				return "return 0";
 		return "''";
 	}
 	/**
@@ -583,6 +588,39 @@
             	if(isString)
                     return propertyName + "=\"" + strValue + "\",";
             	else
+            		    if (strValue.contains( "true"))
+            		        return propertyName + "=" + "True" + ",";
+            		    else
+                            return propertyName + "=" + strValue + ",";
+            else
+                return "";
+		}
+	}
+	
+	/**
+	 * PythonHL Attribute utility
+	 * @param propertyName the name of the attribute property
+	 * @param strValue value to be inserted
+	 * @param isString indicates if the value is a string
+	 * @return resulting string
+	 */
+	public String setPipePropertyHL(String propertyName, String strValue, Boolean isString)
+	{
+		if (propertyName.equals("display_level"))
+		{
+			if (strValue != null && ( strValue.equals("EXPERT") || strValue.equals("PyTango.DispLevel.EXPERT")))
+			{
+				return "display_level=DispLevel.EXPERT,";
+			}
+			else
+				return "";
+		}
+		else
+		{
+            if (strValue != null && strValue.length() > 0)
+            	if(isString)
+                    return propertyName + "=\"" + strValue + "\",";
+            	else
                     return propertyName + "=" + strValue + ",";
             else
                 return "";
Index: PythonUtils.xtend
===================================================================
--- PythonUtils.xtend	(revision 30002)
+++ PythonUtils.xtend	(working copy)
@@ -39,6 +39,7 @@
 import fr.esrf.tango.pogo.pogoDsl.Attribute
 import fr.esrf.tango.pogo.pogoDsl.PogoDeviceClass
 import fr.esrf.tango.pogo.pogoDsl.Property
+import fr.esrf.tango.pogo.pogoDsl.Pipe
 import com.google.inject.Inject
 import static extension fr.esrf.tango.pogo.generator.python.PythonTypeDefinitions.*
 import static extension fr.esrf.tango.pogo.generator.common.StringUtils.*
@@ -77,7 +78,7 @@
     	{
 	    	if (str.contains(","))
 	    	{
-	    		return "[\"" + str.replaceAll(",","\",\"") + "\"]";
+	    		return "[\"" + str.substring(1, str.length()-1).replaceAll(", ",",").replaceAll(",","\", \"") + "\"]";
 	    	}
 	    	else
 	    	{
@@ -204,7 +205,8 @@
             attr.eventCriteria!=null           		||
             attr.getAttType().equals("Spectrum")	||
             attr.getAttType().equals("Image")		||
-            attr.eventCriteria!=null);
+            attr.eventCriteria!=null)				||
+            (attr.enumLabels != null && attr.enumLabels.size > 0);
     }
     
     def hasCmdPropertySet(Command cmd){
@@ -215,6 +217,15 @@
         return (!cmd.argin.type.voidType || !cmd.argout.type.voidType );
     }
     
+    def hasPipePropertySetHL(Pipe pip){
+        return (
+            (pip.label != null && !pip.label.empty) 		||
+            (pip.description != null && !pip.description.empty ) 		||
+            pip.displayLevel.equals("EXPERT") 		||
+            (pip.rwType.contains("WRITE"))
+            )
+    }
+    
     def commandExecution(PogoDeviceClass cls, Command cmd) '''
 		def «cmd.methodName»(self«IF !cmd.argin.type.voidType», argin«ENDIF»):
 		    """ «cmd.description»
@@ -291,7 +302,18 @@
                 «IF cls.description.filestogenerate.toLowerCase.contains("protected regions")»«protectedAreaHL(cls, attribute.name + "_read", "return " + attribute.defaultValueHL, false)»«ELSE»return «attribute.defaultValueDim»«ENDIF»
 
     '''
+      
+    def readPipeMethodHL(PogoDeviceClass cls, Pipe pip) '''
+        def read_«pip.name»(self):
+                «IF cls.description.filestogenerate.toLowerCase.contains("protected regions")»«protectedAreaHL(cls, pip.name + "_read", "return dict(x=0,y=0)", false)»«ELSE»return dict(x=0,y=0)»«ENDIF»
+
+    '''
     
+    def writePipeMethodHL(PogoDeviceClass cls, Pipe pip) '''
+        def write_«pip.name»(self, value):
+                «IF cls.description.filestogenerate.toLowerCase.contains("protected regions")»«protectedAreaHL(cls, pip.name + "_write","print dict(value)", false)»«ELSE»print dict(value)»«ENDIF»
+
+    '''
     def attributeMethodStateMachine(PogoDeviceClass cls, Attribute attribute) '''
 		def is_«attribute.name»_allowed(self, attr):
 		    self.debug_stream("In is_«attribute.name»_allowed()")
@@ -330,6 +352,34 @@
     		}
     	}
     }
+    
+    def readWriteStateMachinePipe(PogoDeviceClass cls, Pipe pip)
+    {
+    	if (!pip.readExcludedStates.empty)
+    	{
+    		if (!pip.writeExcludedStates.empty)
+    		{
+    			return "if attr==attr.READ_REQ:\n    return " + pip.readExcludedStates.ifContentFromListPythonHL + "\nelse:\n    return " + pip.writeExcludedStates.ifContentFromListPythonHL;
+    		}
+    		else
+    		{
+    			return "if attr==attr.READ_REQ:\n    return " + pip.readExcludedStates.ifContentFromListPythonHL + "\nelse:\n    return True";
+    		}
+    		
+    	}
+    	else
+    	{
+    		
+    		if (!pip.writeExcludedStates.empty)
+    		{
+    			return "if attr==attr.READ_REQ:\n    return True\nelse:\n    return " + pip.writeExcludedStates.ifContentFromListPythonHL;
+    		}
+    		else
+    		{
+    			return "if attr==attr.READ_REQ:\n    return True\nelse:\n    return True";
+    		}
+    	}
+    }
 
     def attributeMethodStateMachineHL(PogoDeviceClass cls, Attribute attribute) '''
 		def is_«attribute.name»_allowed(self, attr):
@@ -338,21 +388,28 @@
         «IF attribute.rwType.equals("READ_WRITE") || attribute.rwType.equals("READ_WITH_WRITE")»«IF cls.description.filestogenerate.toLowerCase.contains("protected regions")»«protectedAreaHL(cls, "is_" + attribute.name + "_allowed",cls.readWriteStateMachine(attribute), false)»«ELSE»«cls.readWriteStateMachine(attribute)»«ENDIF»«ENDIF»
 
     '''
+    def pipeMethodStateMachineHL(PogoDeviceClass cls, Pipe pip) '''
+		def is_«pip.name»_allowed(self, attr):
+        «IF pip.rwType.equals("READ")»«IF cls.description.filestogenerate.toLowerCase.contains("protected regions")»«protectedAreaHL(cls, "is_" + pip.name + "_allowed","return " + pip.readExcludedStates.ifContentFromListPythonHL, false)»«ELSE»return «pip.readExcludedStates.ifContentFromListPythonHL»«ENDIF»«ENDIF»
+        «IF pip.rwType.equals("WRITE")»«IF cls.description.filestogenerate.toLowerCase.contains("protected regions")»«protectedAreaHL(cls, "is_" + pip.name + "_allowed","return " + pip.writeExcludedStates.ifContentFromListPythonHL, false)»«ELSE»return «pip.writeExcludedStates.ifContentFromListPythonHL»«ENDIF»«ENDIF»
+        «IF pip.rwType.equals("READ_WRITE") || pip.rwType.equals("READ_WITH_WRITE")»«IF cls.description.filestogenerate.toLowerCase.contains("protected regions")»«protectedAreaHL(cls, "is_" + pip.name + "_allowed",cls.readWriteStateMachinePipe(pip), false)»«ELSE»«cls.readWriteStateMachinePipe(pip)»«ENDIF»«ENDIF»
+
+    '''
     def pythonPropertyClass(Property prop) '''        '«prop.name»':
             [«prop.type.pythonPropType», 
             «IF !prop.description.empty»"«prop.description.oneLineString»"«ELSE» ''«ENDIF»«IF !prop.defaultPropValue.empty»,
-            «IF prop.type.pythonPropType.equals("PyTango.DevString")»["«prop.defaultPropValue.get(0)»"] «ELSEIF prop.type.pythonPropType.equals("PyTango.DevVarStringArray")»[«prop.defaultPropValue.get(0).stringListToStringArray»]«ELSE»«prop.defaultPropValue»«ENDIF»«ELSE»,
+            «IF prop.type.pythonPropType.equals("PyTango.DevString")»["«prop.defaultPropValue.get(0)»"] «ELSEIF prop.type.pythonPropType.equals("PyTango.DevVarStringArray")»«prop.defaultPropValue.toString.stringListToStringArray»«ELSE»«prop.defaultPropValue»«ENDIF»«ELSE»,
             [] «ENDIF»],
     '''
     
     def pythonPropertyClassHL(Property prop) '''
         «prop.name» = class_property(
-                dtype=«prop.type.pythonPropTypeHL»,«IF !prop.defaultPropValue.empty» default_value=«IF prop.type.pythonPropType.equals("PyTango.DevString")»"«prop.defaultPropValue.get(0)»"«ELSEIF prop.type.pythonPropType.equals("PyTango.DevVarStringArray")»«prop.defaultPropValue.get(0).stringListToStringArray»«ELSE»«prop.defaultPropValue.get(0).stringToPyth»«ENDIF»«ENDIF»
+                dtype=«prop.type.pythonPropTypeHL»,«IF !prop.defaultPropValue.empty» default_value=«IF prop.type.pythonPropType.equals("PyTango.DevString")»"«prop.defaultPropValue.get(0)»"«ELSEIF prop.type.pythonPropType.equals("PyTango.DevVarStringArray")»«prop.defaultPropValue.toString.stringListToStringArray»«ELSE»«prop.defaultPropValue.get(0).stringToPyth»«ENDIF»«ENDIF»
             )
     '''
     def pythonPropertyDeviceHL(Property prop) '''
         «prop.name» = device_property(
-                dtype=«prop.type.pythonPropTypeHL»,«IF !prop.defaultPropValue.empty» default_value=«IF prop.type.pythonPropType.equals("PyTango.DevString")»"«prop.defaultPropValue.get(0)»"«ELSEIF prop.type.pythonPropType.equals("PyTango.DevVarStringArray")»«prop.defaultPropValue.get(0).stringListToStringArray»«ELSE»«prop.defaultPropValue.get(0).stringToPyth»«ENDIF»«ENDIF»
+                dtype=«prop.type.pythonPropTypeHL»,«IF !prop.defaultPropValue.empty» default_value=«IF prop.type.pythonPropType.equals("PyTango.DevString")»"«prop.defaultPropValue.get(0)»"«ELSEIF prop.type.pythonPropType.equals("PyTango.DevVarStringArray")»«prop.defaultPropValue.toString.stringListToStringArray»«ELSE»«prop.defaultPropValue.get(0).stringToPyth»«ENDIF»«ENDIF»
             )
     '''
     
@@ -445,11 +502,23 @@
         «setAttrPropertyHL("min_alarm", attr.properties.minAlarm, false)»
         «setAttrPropertyHL("max_warning", attr.properties.maxWarning, false)»
         «setAttrPropertyHL("min_warning", attr.properties.minWarning, false)»
+        «setAttrPropertyHL("memorized", attr.memorized, false)»
+        «setAttrPropertyHL("hw_memorized", attr.memorizedAtInit, false)»
         «setAttrPropertyHL("delta_t", attr.properties.deltaTime, false)»
         «setAttrPropertyHL("delta_val", attr.properties.deltaValue, false)»
-        «setAttrPropertyHL("doc", attr.properties.description.oneLineString, true)»«ELSE»«ENDIF»
+        «setAttrPropertyHL("doc", attr.properties.description.oneLineString, true)»
+        «IF attr.enumLabels!=null»«IF attr.enumLabels.size >0»«setAttrPropertyHL("enum_labels", attr.pythonPipeEnum,  false)»«ENDIF»«ENDIF»«ENDIF»
     )
     '''
+    def pythonPipeClassHL(Pipe pip) '''
+«pip.name» = pipe(
+«IF pip.hasPipePropertySetHL»        «setPipePropertyHL("label", pip.label, true)»
+        «setPipePropertyHL("display_level", pip.displayLevel, false)»
+        «setPipePropertyHL("doc", pip.description.oneLineString, true)»
+        «IF pip.rwType.contains("WRITE")»access=PipeWriteType.PIPE_READ_WRITE«ENDIF»
+«ENDIF»
+    )
+    '''
     
     //======================================================
     def setEventCriteria(PogoDeviceClass cls) '''
Index: PyUtils.java
===================================================================
--- PyUtils.java	(revision 30002)
+++ PyUtils.java	(working copy)
@@ -156,4 +156,28 @@
 			}
 		}
 	}
+
+	//===========================================================
+	/**
+	 * Returns the enum labels on one line 
+	 * @param command the specified command
+	 * @return  the method execution name for specified command
+	 */
+	//===========================================================
+    String pythonPipeEnum(Attribute attr){
+    	if (attr.getEnumLabels() != null)
+    	{
+			StringBuilder	sb = new StringBuilder("[");
+			for (String label : attr.getEnumLabels())
+			{
+				sb.append("\"").append(label).append("\", ");
+			}
+			sb.append("]");
+			return sb.toString();
+    	}
+    	else
+    	{
+    		return "";
+    	}
+    }
 }
